Patch from MiniMyth2 project

Developed-by: Piotr Oniszczuk <piotr.oniszczuk@gmail.com>
Signed-off-by: Piotr Oniszczuk <piotr.oniszczuk@gmail.com>





this reverts https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/drivers/gpio?h=v6.19-rc8&id=471e998c0e31206ff0eac7202b2659698cf9b46e
to fix oops like this:

[   85.260098] ------------[ cut here ]------------
[   85.260102] WARNING: drivers/gpio/gpiolib.c:424 at gpiochip_get_direction+0x28/0x40, CPU#2: cat/7100
[   85.260114] Modules linked in: ip6table_filter ip6_tables iptable_filter ip_tables x_tables mousedev rpcsec_gss_krb5 auth_rpcgss nfsv4 nfs lockd grace nfs_localio sunrpc netfs squashfs openvfd zram ipv6 nls_ascii nls_cp437 sprdwl_ng uwe5622_bsp_sdio cfg80211 sg uinput snd_soc_hdmi_codec hci_uart btqca btrtl btbcm sun50iw9_codec btintel snd_soc_simple_card snd_soc_spdif_tx snd_soc_simple_card_utils ir_nec_decoder snd_soc_sunxi_ahub bluetooth snd_soc_sunxi_machine dw_hdmi_i2s_audio dw_hdmi_cec rc_beelink_gs1 sun4i_spdif sun4i_codec snd_soc_sunxi_ahub_dam sunxi_cir snd_soc_core snd_compress snd_pcm_dmaengine ecdh_generic snd_pcm sunxi_cedrus(C) pwrseq_core snd_timer videobuf2_dma_contig rfkill ecc snd dwmac_sun8i v4l2_mem2mem soundcore sun8i_thermal videobuf2_v4l2 sunxi_wdt videodev videobuf2_memops videobuf2_common governor_performance mc panfrost gpu_sched cpufreq_dt overlay
[   85.260314] CPU: 2 UID: 0 PID: 7100 Comm: cat Tainted: G        WC          6.19.0-rc8 #3 PREEMPT  c793c7c73504469ad4472a202c09563fafe1a232
[   85.260325] Tainted: [W]=WARN, [C]=CRAP
[   85.260328] Hardware name: OrangePi Zero3 (DT)
[   85.260332] pstate: 60400005 (nZCv daif +PAN -UAO -TCO -DIT -SSBS BTYPE=--)
[   85.260339] pc : gpiochip_get_direction+0x28/0x40
[   85.260348] lr : gpiod_get_direction+0x90/0x1a0
[   85.260355] sp : ffff8000841dbad0
[   85.260358] x29: ffff8000841dbaf0 x28: ffff0000c1089f20 x27: ffff80008158aaa8
[   85.260371] x26: 0000000000000000 x25: 00000000000000a6 x24: ffff800081506fa0
[   85.260384] x23: 0000000000000001 x22: 0000000000000000 x21: ffff0000c16dcc18
[   85.260396] x20: ffff0000c16dc800 x19: ffff0000c1089f20 x18: 000000000000000a
[   85.260409] x17: 0000000000000000 x16: 0000000000000000 x15: 0000000000000000
[   85.260421] x14: 0000000000000000 x13: 00000000ffff0a02 x12: 0000000000000003
[   85.260433] x11: ffff0000d780610d x10: 000000000000000a x9 : 0000000000000000
[   85.260445] x8 : ffff0000d7647000 x7 : 0000000000000138 x6 : 0000000000000000
[   85.260458] x5 : ffff800081884a40 x4 : 0000000000000001 x3 : aaaaaaaaaaaaaaab
[   85.260470] x2 : 0000000000000000 x1 : 00000000000000a6 x0 : ffff0000c16d9c80
[   85.260483] Call trace:
[   85.260486]  gpiochip_get_direction+0x28/0x40 (P)
[   85.260496]  gpiolib_dbg_show+0x12c/0x278
[   85.260504]  gpiolib_seq_show+0x180/0x198
[   85.260512]  seq_read_iter+0xf8/0x460
[   85.260521]  seq_read+0xdc/0x120
[   85.260529]  full_proxy_read+0x60/0xa0
[   85.260537]  vfs_read+0xc0/0x31c
[   85.260546]  ksys_read+0x5c/0xf4
[   85.260555]  __arm64_sys_read+0x14/0x20
[   85.260563]  el0_svc_common.constprop.0+0x58/0x124
[   85.260573]  do_el0_svc+0x18/0x20
[   85.260581]  el0_svc+0x30/0xf8
[   85.260589]  el0t_64_sync_handler+0x98/0xe0
[   85.260597]  el0t_64_sync+0x19c/0x1a0
[   85.260605] ---[ end trace 0000000000000000 ]---



diff --speed-large-files --no-dereference --minimal -Naur linux-6.19-rc8/drivers/gpio/gpiolib.c linux-6.19-rc8/drivers/gpio/gpiolib.c
--- linux-6.19-rc8/drivers/gpio/gpiolib.c	2026-02-01 23:01:13.000000000 +0100
+++ linux-6.19-rc8/drivers/gpio/gpiolib.c	2026-02-05 21:30:08.414668861 +0100
@@ -468,6 +468,9 @@
 	    test_bit(GPIOD_FLAG_IS_OUT, &flags))
 		return 0;
 
+	if (!guard.gc->get_direction)
+		return -ENOTSUPP;
+
 	ret = gpiochip_get_direction(guard.gc, offset);
 	if (ret < 0)
 		return ret;
--

MiniMyth2

